File: ex5/interfaces.pas
{$MODE OBJFPC}
{$H+}

program interfaces;

uses
  SysUtils;

type
  IProduct = interface
    function GetName: string;
    function GetPrice: double;
    property Name: string read GetName;
    property Price: double read GetPrice;
  end;

  TProduct = class(TInterfacedObject, IProduct)
  private
    FName: string;
    FPrice: double;
  public
    constructor Create(const AName: string; const APrice: double);
    function GetName: string;
    function GetPrice: double;
  end;

  TShoppingCart = class
  private
    FItems: array of IProduct;
    function GetCount: integer;
  public
    constructor Create;
    destructor Destroy; override;
    procedure AddItem(const AProduct: IProduct);
    function GetTotalCost: double;
    property Count: integer read GetCount;
  end;

  { TProduct }

  constructor TProduct.Create(const AName: string; const APrice: double);
  begin
    FName := AName;
    FPrice := APrice;
  end;

  function TProduct.GetName: string;
  begin
    Result := FName;
  end;

  function TProduct.GetPrice: double;
  begin
    Result := FPrice;
  end;

  { TShoppingCart }

  constructor TShoppingCart.Create;
  begin
    SetLength(FItems, 0);
  end;

  destructor TShoppingCart.Destroy;
  var
    I: integer;
  begin
    for I := 0 to High(FItems) do
      FItems[I] := nil;
    inherited;
  end;

  function TShoppingCart.GetCount: integer;
  begin
    Result := Length(FItems);
  end;

  procedure TShoppingCart.AddItem(const AProduct: IProduct);
  begin
    SetLength(FItems, Length(FItems) + 1);
    FItems[High(FItems)] := AProduct;
  end;

  function TShoppingCart.GetTotalCost: double;
  var
    I: integer;
    TotalCost: double;
  begin
    TotalCost := 0;
    for I := 0 to High(FItems) do
      TotalCost := TotalCost + FItems[I].Price;
    Result := TotalCost;
  end;

var
  Cart: TShoppingCart;
  ProductName: string;
  ProductPrice: double;
  Choice: string;
  I: integer;

begin
  Cart := TShoppingCart.Create;
  try
    repeat
      Write('Enter product name (or "x" to finish): ');
      ReadLn(ProductName);

      if ProductName.ToLower <> 'x' then
      begin
        Write('Enter product price: ');
        ReadLn(ProductPrice);

        Cart.AddItem(TProduct.Create(ProductName, ProductPrice));
      end;

    until ProductName.ToLower = 'x';

    WriteLn('Shopping Cart Contents:');
    WriteLn('------------------------');
    for I := 0 to Cart.Count - 1 do
      WriteLn(Cart.FItems[I].Name + ': $' + Cart.FItems[I].Price.ToString);

    WriteLn('------------------------');
    WriteLn('Total Cost: $' + Cart.GetTotalCost.ToString);

    Write('Press Enter to exit...');
    ReadLn(Choice);
  finally
    Cart.Free;
  end;
end.

File: C#-Translated/Interfaces/Transpiler/Program/Interfaces.cs
namespace interfaces
{
    using SysUtils;

    public static class Program
    {
        private static TShoppingCart Cart;
        private static string ProductName;
        private static double ProductPrice;
        private static string Choice;
        private static int I;
        public static void Main(string[] args)
        {
            Cart = TShoppingCart.Create;
            try
            {
                do
                {
                    Write("Enter product name (or \"x\" to finish): ");
                    ReadLn(ProductName);
                    if (ProductName.ToLower != "x")
                    {
                        Write("Enter product price: ");
                        ReadLn(ProductPrice);
                        Cart.AddItem(TProduct.Create(ProductName, ProductPrice));
                    }
                }
                while (ProductName.ToLower == "x");
                WriteLn("Shopping Cart Contents:");
                WriteLn("------------------------");
                for (int i = 0; i < Cart.Count - 1; i++)
                    WriteLn(Cart.FItems[I]() + ": $" + Cart.FItems[I]()());
                WriteLn("------------------------");
                WriteLn("Total Cost: $" + Cart.GetTotalCost.ToString);
                Write("Press Enter to exit...");
                ReadLn(Choice);
            }
            finally
            {
                Cart.Free;
            }
        }

        interface IProduct
        {
            string GetName()
            {
                string Result;
                return Result;
            }

            double GetPrice()
            {
                double Result;
                return Result;
            }

            string<> Name double<> Price
        }

        private class TProduct : TInterfacedObject, IProduct
        {
            private string FName;
            private double FPrice;
            public TProduct(string AName, double APrice)
            {
                FName = AName;
                FPrice = APrice;
            }

            public string GetName()
            {
                string Result;
                Result = FName;
                return Result;
            }

            public double GetPrice()
            {
                double Result;
                Result = FPrice;
                return Result;
            }
        }

        private class TShoppingCart
        {
            integer<> Count private IProduct[] FItems;
            private int GetCount()
            {
                int Result;
                Result = Length(FItems);
                return Result;
            }

            public TShoppingCart()
            {
                SetLength(FItems, 0);
            }

            override public ~TShoppingCart()
            {
                for (int i = 0; i < High(FItems); i++)
                    FItems[I] = null;
                base;
            }

            public void AddItem(IProduct AProduct)
            {
                SetLength(FItems, Length(FItems) + 1);
                FItems[High(FItems)] = AProduct;
            }

            public double GetTotalCost()
            {
                double Result;
                TotalCost = 0;
                for (int i = 0; i < High(FItems); i++)
                    TotalCost = TotalCost + FItems[I]();
                Result = TotalCost;
                return Result;
            }
        }
    }
}